addon/components/step-manager.js

```diff
@@ -189,25 +232,34 @@ export default Component.extend({
      */
     'transition-to-step'(to, value) {
       const from = get(this, 'transitions.currentStep');
+      const validator = this['will-transition'];
 
-      if (
-        this['will-transition'] &&
-        this['will-transition']({ value, from, to }) === false
-      ) {
+      // Prevent multiple transitions
+      if (get(this, 'loading')) {
         return;
       }
 
-      // Update the `currentStep` if it's mutable
-      if (this.attrs.currentStep && this.attrs.currentStep.update) {
-        this.attrs.currentStep.update(to);
-      }
+      if (validator && typeof validator === 'function') {
+        const result = validator({ value, from, to });
 
-      // Activate the next step
-      get(this, 'transitions').activate(to);
-
-      if (this['did-transition']) {
-        this['did-transition']({ value, from, to });
+        if (result) {
+          if (typeof result.then === 'function') {
```

Comment:
Kind of a nit-pick, but would you mind updating this and the condition check above to be a single conditional instead? Just to prevent the nested `if`-statements here.

Reply:
Yeah good comment, I will fix that.